### PR TITLE
Ensure AddPostStartHook always succeeds, or die

### LINKED ISSUE #15819: RBAC Migration Followup #7: Post start hook build error should fatal server
`glog.Fatal` on post stark hook error

### PR DESCRIPTION
**Pull request by simo5**
Fixes #15819

### CODE REVIEW

**Comment by simo5** [comment]
@enj PTAL

**Comment by simo5** [comment]
Flake #15874

**Comment by simo5** [comment]
Bug for the ansible flake:
https://github.com/openshift/openshift-ansible/issues/5146

**Comment by sdodson** [comment]
/retest
https://github.com/openshift/aos-cd-jobs/pull/578 should've fixed ci jobs

**Comment by simo5** [comment]
@sdodson it's fixed, thanks

**Comment by simo5** [comment]
@deads2k PTAL

**Comment by enj** [comment]
This LGTM, but I will wait for David to have a look.

File: pkg/cmd/server/origin/openshift_apiserver.go [hunk 1]
```diff
@@ -342,12 +342,12 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
...
-	s.GenericAPIServer.AddPostStartHook("project.openshift.io-projectcache", c.startProjectCache)
-	s.GenericAPIServer.AddPostStartHook("project.openshift.io-projectauthorizationcache", c.startProjectAuthorizationCache)
-	s.GenericAPIServer.AddPostStartHook("security.openshift.io-bootstrapscc", c.bootstrapSCC)
-	s.GenericAPIServer.AddPostStartHook("authorization.openshift.io-ensureSARolesDefault", c.ensureDefaultNamespaceServiceAccountRoles)
-	s.GenericAPIServer.AddPostStartHook("authorization.openshift.io-ensureopenshift-infra", c.ensureOpenShiftInfraNamespace)
+	ensureAddPostStartHook(s.GenericAPIServer, "quota.openshift.io-clusterquotamapping", c.startClusterQuotaMapping)
+	ensureAddPostStartHook(s.GenericAPIServer, "project.openshift.io-projectcache", c.startProjectCache)
+	ensureAddPostStartHook(s.GenericAPIServer, "project.openshift.io-projectauthorizationcache", c.startProjectAuthorizationCache)
+	ensureAddPostStartHook(s.GenericAPIServer, "security.openshift.io-bootstrapscc", c.bootstrapSCC)
+	ensureAddPostStartHook(s.GenericAPIServer, "authorization.openshift.io-ensureSARolesDefault", c.ensureDefaultNamespaceServiceAccountRoles)
```
Comment (smarterclayton) [inline pkg/cmd/server/origin/openshift_apiserver.go]: Drive by - this casing is inconsistent and should be fixed

Comment (simo5) [inline pkg/cmd/server/origin/openshift_apiserver.go]: Sorry fail to see the inconsistency, ensure is lower case because it is a private function.

**Comment by smarterclayton** [comment]
"authorization.openshift.io-ensureSARolesDefault" is inconsistent with
every other hook name because it has upper case letters.  If we have a
pattern, we should follow it.

**Comment by enj** [comment]
> "authorization.openshift.io-ensureSARolesDefault" is inconsistent with every other hook name because it has upper case letters. If we have a pattern, we should follow it.

@deads2k what would like to change it to since you named it as such?

**Comment by deads2k** [comment]
> @deads2k what would like to change it to since you named it as such

I'm biased towards lower case.  Notice that the groups match, then a dash, then a camelCase name.

File: pkg/cmd/server/origin/ensure.go [hunk 2]
```diff
@@ -22,3 +22,11 @@ func (c *MasterConfig) ensureOpenShiftSharedResourcesNamespace(context genericap
 	}
 	return nil
 }
+
+// TODO: We should have a single place where we register all post start hook
+// and make this function unnecessary
+func ensureAddPostStartHook(genericAPIServer *genericapiserver.GenericAPIServer, name string, hook genericapiserver.PostStartHookFunc) {
```
Comment (deads2k) [inline pkg/cmd/server/origin/ensure.go]: Move this upstream.  `AddPostStartHookOrDie`.  We want the separate function with an error to allow testing of the error condition (collision on names as I recall).

File: pkg/cmd/server/origin/openshift_apiserver.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/origin/openshift_apiserver.go]: I could see camel-casing in this direction `security.openshift.io-BootstrapSCC`.  The group matches the group, allows blind composition later, and sorts well.

File: pkg/cmd/server/origin/ensure.go (hunk 2 above)
Comment (simo5) [inline pkg/cmd/server/origin/ensure.go]: @deads2k upstream doesn't really need it as they run all their AddPostStartHook() calls into a function that already fatals if an error is returned. So it seem like upstreaming an OrDie version there would be useless for upstream.

That's why we have a TODO to create also a registration function in origin as well that properly deal with it at some point.

**Comment by simo5** [comment]
@smarterclayton @deads2k for the casing, ok if we open a separate issue ?
I am not really touching those functions here, just changing a user, so the fact those names appear in this PR is just incidental.

File: pkg/cmd/server/origin/ensure.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/origin/ensure.go]: > @deads2k upstream doesn't really need it as they run all their AddPostStartHook() calls into a function that already fatals if an error is returned. So it seem like upstreaming an OrDie version there would be useless for upstream.

I don't think they'll object to having an `OrDie` variant and it makes out dependency set cleaner as collapse onto upstream.

**Comment by deads2k** [comment]
> @smarterclayton @deads2k for the casing, ok if we open a separate issue ?
> I am not really touching those functions here, just changing a user, so the fact those names appear in this PR is just incidental.

I don't really care.  The names aren't special.

**Comment by simo5** [comment]
Ok I am opening a followup issue for the casing then, given that:
- this PR is not really touching names
- there seem no consensus on what they should be changed to anyway

**Comment by simo5** [comment]
Ok move that discussion here #15926 please

File: pkg/cmd/server/origin/ensure.go (hunk 2 above)
Comment (simo5) [inline pkg/cmd/server/origin/ensure.go]: ok

**Comment by simo5** [comment]
Updated @deads2k PTAL

**Comment by simo5** [comment]
Flaked on #14575

File: vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/hooks.go [hunk 3]
```diff
@@ -90,6 +90,15 @@ func (s *GenericAPIServer) AddPostStartHook(name string, hook PostStartHookFunc)
 	return nil
 }
 
+// AddPostStartHookOrDie allows you to add a PostStartHook, but dies on failure
+func (s *GenericAPIServer) AddPostStartHookOrDie(name string, hook PostStartHookFunc) error {
+	err := s.AddPostStartHook(name, hook)
+	if err != nil {
```
Comment (enj) [inline vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/hooks.go]: `if err := s.AddPostStartHook(name, hook); err != nil {`

It is nice to have variables scoped to an if statement.

**Pull request merged**